tests/unit/legate/util/test_system.py

```diff
@@ -105,3 +105,77 @@ def test_gpus_osx(self) -> None:
         msg = "GPU execution is not available on OSX."
         with pytest.raises(RuntimeError, match=msg):
             s.gpus
+
+
+class Test_expand_range:
+    def test_errors(self) -> None:
+        with pytest.raises(ValueError):
+            m.expand_range("foo")
+
+    def test_empty(self) -> None:
+        assert m.expand_range("") == ()
+
+    @pytest.mark.parametrize("val", ("0", "1", "12", "100"))
+    def test_single_number(self, val: str) -> None:
+        assert m.expand_range(val) == (int(val),)
+
+    @pytest.mark.parametrize("val", ("0-10", "1-2", "12-25"))
+    def test_range(self, val: str) -> None:
+        start, stop = val.split("-")
+        assert m.expand_range(val) == tuple(range(int(start), int(stop) + 1))
+
+
+class Test_extract_values:
+    def test_errors(self) -> None:
+        with pytest.raises(ValueError):
+            m.extract_values("foo")
+
+    def test_empty(self) -> None:
+        assert m.extract_values("") == ()
+
+    @pytest.mark.parametrize("val", ("0", "1,2", "3,5,7"))
+    def test_individual(self, val: str) -> None:
+        expected = {
+            "0": (0,),
+            "1,2": (1, 2),
+            "3,5,7": (3, 5, 7),
+        }
+        assert m.extract_values(val) == expected[val]
+
+    @pytest.mark.parametrize("val", ("2,1", "8,5,3,2", "1,3,2,5,4,7,6"))
+    def test_individual_ordered(self, val: str) -> None:
+        expected = {
+            "2,1": (1, 2),
+            "8,5,3,2": (2, 3, 5, 8),
+            "1,3,2,5,4,7,6": (1, 2, 3, 4, 5, 6, 7),
+        }
+        assert m.extract_values(val) == expected[val]
+
+    @pytest.mark.parametrize("val", ("0-2", "0-2,4-5", "0-1,3-5,8-11"))
+    def test_range(self, val: str) -> None:
+        expected = {
+            "0-2": (0, 1, 2),
+            "0-2,4-5": (0, 1, 2, 4, 5),
+            "0-1,3-5,8-11": (0, 1, 3, 4, 5, 8, 9, 10, 11),
+        }
+        assert m.extract_values(val) == expected[val]
+
+    @pytest.mark.parametrize("val", ("2-3,0-1", "0-1,4-5,2-3"))
+    def test_range_ordered(self, val: str) -> None:
+        expected = {
+            "2-3,0-1": (0, 1, 2, 3),
+            "0-1,4-5,2-3": (0, 1, 2, 3, 4, 5),
+        }
+        assert m.extract_values(val) == expected[val]
+
+    @pytest.mark.parametrize(
+        "val", ("0,1-2", "1-2,0", "0,1-2,3,4-5,6", "5-6,4,1-3,0")
+    )
+    def test_mixed(self, val: str) -> None:
+        expected = {
+            "0,1-2": (0, 1, 2),
+            "1-2,0": (0, 1, 2),
+            "0,1-2,3,4-5,6": (0, 1, 2, 3, 4, 5, 6),
+            "5-6,4,1-3,0": (0, 1, 2, 3, 4, 5, 6),
+        }
```

Comment:
BTW just FYI it's common to move values + expected values together at the same time through parameters. The danger here is that the keys in the `expected` dict become out of sync with the `val` parameter values since they are duplicated. I think this is fine in this small case, but if you are interested there are more examples, etc on the pytest site

Reply:
Created PR https://github.com/nv-legate/legate.core/pull/851 because I really like the non-duplicating version.